miner/collator.go

```diff
@@ -0,0 +1,402 @@
+				if atomic.CompareAndSwapInt32(bs.interruptHandled, interruptNotHandled, interruptIsHandled) {
+					var ratio float64 = 0.1
+					bs.worker.resubmitAdjustCh <- &intervalAdjust{
+						ratio: ratio,
+						inc:   true,
+					}
+				}
+				return ErrInterruptRecommit, nil
+			} else {
+				return ErrInterruptNewHead, nil
+			}
+		}
+
+		if bs.env.gasPool.Gas() < params.TxGas {
+			return ErrGasLimitReached, nil
+		}
+
+		// Check whether the tx is replay protected. If we're not in the EIP155 hf
+		// phase, start ignoring the sender until we do.
+		if tx.Protected() && !bs.worker.chainConfig.IsEIP155(bs.env.header.Number) {
+			return ErrTxTypeNotSupported, nil
+		}
+
+		// TODO can this error also be returned by commitTransaction below?
+		_, err := tx.EffectiveGasTip(bs.env.header.BaseFee)
+		if err != nil {
+			return ErrGasFeeCapTooLow, nil
+		}
+
+		snapshot := bs.env.state.Snapshot()
+		bs.env.state.Prepare(tx.Hash(), bs.env.tcount+tcount)
+		txLogs, err := bs.worker.commitTransaction(bs.env, tx, bs.env.etherbase)
+		if err != nil {
+			switch {
+			case errors.Is(err, core.ErrGasLimitReached):
+				// this should never be reached.
+				// should be caught above
+				retErr = ErrGasLimitReached
+			case errors.Is(err, core.ErrNonceTooLow):
+				retErr = ErrNonceTooLow
+			case errors.Is(err, core.ErrNonceTooHigh):
+				retErr = ErrNonceTooHigh
+			case errors.Is(err, core.ErrTxTypeNotSupported):
+				// TODO check that this unspported tx type is the same as the one caught above
+				retErr = ErrTxTypeNotSupported
+			default:
+				retErr = ErrStrange
+			}
+
+			bs.logs = bs.logs[:len(bs.logs)-tcount]
+			bs.env.state.RevertToSnapshot(bs.snapshots[len(bs.snapshots)-tcount])
+			bs.snapshots = bs.snapshots[:len(bs.snapshots)-tcount]
+
+			return retErr, nil
+		} else {
+			bs.logs = append(bs.logs, txLogs...)
+			bs.snapshots = append(bs.snapshots, snapshot)
+			tcount++
+		}
+	}
+
+	retReceipts := bs.env.receipts[bs.env.tcount:]
+	bs.env.tcount += tcount
+
+	return nil, retReceipts
+}
+
+func (bs *blockState) State() vm.StateReader {
+	return bs.env.state
+}
+
+func (bs *blockState) Signer() types.Signer {
+	return bs.env.signer
+}
+
+// TODO change return from bool to error
+func (bs *blockState) Commit() bool {
+	if bs.committed {
+		return false
+	}
+
+	if bs.interrupt != nil && atomic.LoadInt32(bs.interrupt) != commitInterruptNone {
+		if atomic.CompareAndSwapInt32(bs.interruptHandled, interruptNotHandled, interruptIsHandled) && atomic.LoadInt32(bs.interrupt) == commitInterruptResubmit {
+			// Notify resubmit loop to increase resubmitting interval due to too frequent commits.
+			var ratio float64 = 0.1
+			bs.worker.resubmitAdjustCh <- &intervalAdjust{
+				ratio: ratio,
+				inc:   true,
+			}
+		}
+		return false
+	}
+
+	bs.commitMu.Lock()
+	defer bs.commitMu.Unlock()
+	if *bs.done {
+		return false
+	}
+	if bs.shouldSeal {
+		bs.worker.commit(bs.env.copy(), bs.worker.fullTaskHook, true, bs.start)
+	}
```

Comment:
This checking for `done` and based on it calling or not calling `worker.commit` seems out of scope. If you already hav the result here, it's simpler to just always call commit and have that potentially reject it. It stops a lot of API leak across boundaries.